kolibri/core/logger/utils/exam_log_migration.py

```diff
@@ -285,3 +302,31 @@ def migrate_from_exam_logs(source_logs):  # noqa C901
                 _update_attempt_logs(masterylog_id, logs)
         i += BATCH_READ_SIZE
         logs = source_logs[i : i + BATCH_READ_SIZE]
+
+    if unprocessed_attempt_log_ids:
+        examlog_id_to_masterylog = {}
+        unprocessed_attempt_log_ids = list(unprocessed_attempt_log_ids)
+        attempt_logs = []
+        i = 0
+        source_logs = ExamAttemptLog.objects.filter(
+            id__in=unprocessed_attempt_log_ids[i : i + BATCH_READ_SIZE]
+        ).annotate(exam_id=F("examlog__exam_id"))
+        for examattemptlog in source_logs:
+            if examattemptlog.examlog_id not in examlog_id_to_masterylog:
+                examlog_id_to_masterylog[examattemptlog.examlog_id] = (
+                    MasteryLog.objects.filter(
+                        user_id=examattemptlog.user_id,
+                        summarylog__content_id=examattemptlog.exam_id,
+                    )
+                    .values_list("id", flat=True)
+                    .first()
+                )
+            masterylog_id = examlog_id_to_masterylog[examattemptlog.examlog_id]
+            if masterylog_id:
+                attemptlog = _create_attemptlog(examattemptlog, None, masterylog_id)
+                attempt_logs.append(attemptlog)
+
+    for masterylog_id, logs in groupby(
+        sorted(attempt_logs, key=lambda x: x.masterylog_id), lambda x: x.masterylog_id
```

Comment:
This can achieve a `referenced before assignment` error for `attempt_logs` here if there are no are logs of any kind. It's defined in the batch loop but then redefined later with these new updates. It's not clear which variable this for-loop is referencing. Perhaps separate this into a separate function?

Reply:
Fair - can separate it out.

The function definitely is too long.